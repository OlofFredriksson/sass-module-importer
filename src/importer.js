import path from "node:path";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";
import { createRequire } from "node:module";

import resolvePackagePath from "resolve-package-path";
import { exports, legacy } from "resolve.exports";

import { getPackageNameFromPath } from "./parsePackageName";

const require = createRequire(import.meta.url);
const WEBPACK_NODE_MODULE_PREFIX = "~";
export const moduleImporter = {
    findFileUrl(url) {
        let findUrl = url;
        if (url.startsWith(WEBPACK_NODE_MODULE_PREFIX)) {
            findUrl = url.substring(1);
        }

        const packageName = getPackageNameFromPath(findUrl);
        const fileName = findUrl.split(packageName)[1];
        const packagePath = resolvePackagePath(packageName, process.cwd());

        /* Validate if existing package */
        if (!packagePath) {
            return null;
        }

        const packageJson = JSON.parse(
            readFileSync(packagePath, { encoding: "utf-8" }),
        );

        const directory = path.dirname(packagePath);

        /* Check exports */
        try {
            const match = exports(packageJson, fileName.substring(1), {
                conditions: ["sass"],
            });
            if (match && match.length === 1) {
                return new URL(pathToFileURL(path.join(directory, match[0])));
            }
        } catch {
            /* empty */
        }

        /* Check main fields (only applies if only package path is given) */
        if (!fileName) {
            const match = legacy(packageJson, { fields: ["sass", "main"] });
            if (match) {
                return new URL(pathToFileURL(path.join(directory, match)));
            }
        }

        /* Direct link */
        const search = [
            `${fileName}.css`,
            `${fileName}.scss`,
            `_${fileName}.scss`,
            `${fileName}`,
        ];

        for (const variant of search) {
            try {
                const moduleName = path.posix.join(directory, variant);
                const resolved = require.resolve(moduleName);
                return new URL(pathToFileURL(resolved));
            } catch (err) {
                if (err.code !== "MODULE_NOT_FOUND") {
                    throw err;
                }
            }
        }
        return null;
    },
};
