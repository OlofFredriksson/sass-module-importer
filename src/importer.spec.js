import fs from "node:fs";
import process from "node:process";

import { compileString } from "sass";
import { jest } from "@jest/globals";
import { moduleImporter } from "./importer";

function init(mainEntry) {
    const spyProcess = jest.spyOn(process, "cwd");

    spyProcess.mockReturnValue("./fixtures");

    return compileString(fs.readFileSync(`./fixtures/${mainEntry}`, "utf-8"), {
        style: "expanded",
        importers: [moduleImporter],
    }).css;
}

it("should be able to transform scss using package without exports and main fields", () => {
    expect(init("main.scss")).toMatchSnapshot();
});

it("should be able to transform scss using package with exports", () => {
    expect(init("exports.scss")).toMatchSnapshot();
});

it("should be able to transform scss using package with main field", () => {
    expect(init("mainField.scss")).toMatchSnapshot();
});

it("should be able to transform scss using package with sass fields", () => {
    expect(init("sassField.scss")).toMatchSnapshot();
});
