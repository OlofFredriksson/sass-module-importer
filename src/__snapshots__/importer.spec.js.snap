// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to transform scss using package with exports 1`] = `
"span {
  color: pink;
}

body {
  color: orange;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss using package with main field 1`] = `
".main {
  color: blue;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss using package with sass fields 1`] = `
".main {
  color: green;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss using package without exports and main fields 1`] = `
"html,
body {
  background-color: #12c344;
}

.bar {
  color: greenyellow;
}

.foo {
  color: green;
}"
`;
